File: sw/python/model/pendulum_plant.py
import numpy as np


class PendulumPlant:
    def __init__(self, mass=1.0, length=0.5, damping=0.1, gravity=9.81,
                 coulomb_fric=0.0, inertia=None, torque_limit=np.inf):

        """
        The PendulumPlant class contains the kinematics and dynamics
        of the simple pendulum.
        The class is initilized with the following parameters:
            - mass:         float, pendulum mass, unit: kg
            - length:       float, pendulum length, unit: m
            - damping:      float, damping factor (proportional to velocity),
                            unit: kg*m/s
            - gravity:      float, gravity (positive direction points down),
                            unit: m/s^2
            - coulomb_fric: float, friction term, (independent of magnitude
                            of velocity), unit: Nm
            - inertia:      float, inertia of the pendulum
                            (defaults to point mass inertia)
                            unit: kg*m^2
            - torque_limit: float, maximum torque that the motor can apply
                            unit: Nm

        The state of the pendulum in this class is described by
            state = [angle, angular velocity]
            (array like with len(state)=2)
            in units: rad and rad/s
        The zero state of the angle corresponds to the pendulum hanging down.
        The plant expects an actuation input (tau) either as float or array like
        (in which case the first entry is used (which should be a float))
        """

        self.m = mass
        self.l = length
        self.b = damping
        self.g = gravity
        self.coulomb_fric = coulomb_fric
        if inertia is None:
            self.inertia = mass*length*length
        else:
            self.inertia = inertia

        self.torque_limit = torque_limit

        self.dof = 1
        self.n_actuators = 1
        self.base = [0, 0]
        self.n_links = 1
        self.workspace_range = [[-1.2*self.l, 1.2*self.l],
                                [-1.2*self.l, 1.2*self.l]]

    def forward_kinematics(self, pos):

        """
        Computes the forward kinematics.
        input:
            - pos:  float, angle of the pendulum
        returns:
            - A list containing one list (for one end-effector)
              The inner list contains the x and y coordinates
              for the end-effector of the pendulum
        """

        ee_pos_x = self.l * np.sin(pos)
        ee_pos_y = -self.l*np.cos(pos)
        return [[ee_pos_x, ee_pos_y]]

    def inverse_kinematics(self, ee_pos):

        """
        Comutes inverse kinematics
        input:
            - ee_pos: array like, len(state)=2
                      contains the x and y position of the end_effector
                      floats, units: m
        returns:
            - float, angle of the pendulum, unit: rad
        """

        pos = np.arctan2(ee_pos[1], ee_pos[0])
        return pos

    def forward_dynamics(self, state, tau):

        """
        Computes forward dynamics
        input:
            - state: array like, len(state)=2
                     The state of the pendulum [angle, angular velocity]
                     floats, units: rad, rad/s
            - tau: float, motor torque, unit: Nm
        returns:
            - float, angular acceleration, unit: rad/s^2
        """

        torque = np.clip(tau, -np.asarray(self.torque_limit),
                         np.asarray(self.torque_limit))

        accn = (torque - self.m * self.g * self.l * np.sin(state[0]) -
                self.b * state[1] -
                np.sign(state[1])*self.coulomb_fric) / self.inertia
        return accn

    def rhs(self, t, state, tau):

        """
        Computes the integrand of the equations of motion.
        input:
            - t:        float, time, not used
            - state:    array like, len(state)=2
                        The state of the pendulum [angle, angular velocity]
                        floats, units: rad, rad/s
            - tau: float or array like, motor torque, unit: Nm
        returns:
            - integrand, i.e.
              [angular velocity, angular acceleration]
        """

        if isinstance(tau, (list, tuple, np.ndarray)):
            torque = tau[0]
        else:
            torque = tau

        accn = self.forward_dynamics(state, torque)

        res = np.zeros(2*self.dof)
        res[0] = state[1]
        res[1] = accn
        return res

File: software/cpp/model/src/pendulum_plant.cpp
#include <stdlib.h>
#include <iostream>
#include <math.h>

#include "pendulum_plant.hpp"

void PendulumPlant::set_parameters(double m, double l, double d, double cf, double g, double I, double tl){
    mass = m;
    length = l;
    damping = d;
    coulomb_friction = cf;
    gravity = g;
    inertia = I;
    torque_limit = tl;
}

double PendulumPlant::forward_dynamics(double pos, double vel, double tau){
    double fric;
    if (vel > 0.0){
        fric = coulomb_friction;
    }
    else if (vel < 0.0) {
        fric = -1.0*coulomb_friction;
    }
    else {
        fric = 0.0;
    }
    double accn = (tau - mass*gravity*length*sin(pos) - damping*vel - fric) / inertia;
    //std::cout << "plant fd: pos: " << pos << "vel: " << vel << "tau" << tau << "acc: " << accn << std::endl;
    return accn;
}
State PendulumPlant::rhs(double t, double pos, double vel, double tau){
    double accn = forward_dynamics(pos, vel, tau);
    State r;
    r.pos = vel;
    r.vel = accn;
    //std::cout << "plant rhs: (" << r.pos << ", " << r.vel << ")" << std::endl;
    return r;
}

double PendulumPlant::get_torque_limit(){
    return torque_limit;
}

double PendulumPlant::calculate_potential_energy(State x){
    return mass*gravity*length*(1.0 - cos(x.pos));
}

double PendulumPlant::calculate_kinetic_energy(State x){
    return 0.5*mass*pow(length*x.vel, 2);
}

double PendulumPlant::calculate_total_energy(State x){
    return calculate_kinetic_energy(x) + calculate_potential_energy(x);
}
